File: python/408. Valid Word Abbreviation.py
class Solution:
    """
    @param word: a non-empty string
    @param abbr: an abbreviation
    @return: true if string matches with the given abbr or false
    """
    def valid_word_abbreviation(self, word: str, abbr: str) -> bool:
        # write your code here
        i, j = 0, 0
        n, m = len(word), len(abbr)
        while i < n and j < m:
            if abbr[j].isdigit():
                if abbr[j] == '0':
                    return False
                num = 0
                while j < m and abbr[j].isdigit():
                    num = num * 10 + int(abbr[j])
                    j += 1
                i += num
            else:
                if word[i] != abbr[j]:
                    return False
                i += 1
                j += 1
        return i == n and j == m

File: rust/408. Valid Word Abbreviation.rs
impl Solution {
    pub fn valid_word_abbreviation(word: String, abbr: String) -> bool {
        let (word, abbr) = (word.as_bytes(), abbr.as_bytes());
        let (mut i, mut j) = (0, 0);

        while i < word.len() && j < abbr.len() {
            match abbr[j] {
                b'0' => return false,
                b'1'..=b'9' => {
                    let mut num = 0;
                    while j < abbr.len() && abbr[j].is_ascii_digit() {
                        num = num * 10 + (abbr[j] - b'0') as usize;
                        j += 1;
                    }
                    i += num;
                }
                _ => {
                    if word.get(i) != abbr.get(j) {
                        return false;
                    }
                    i += 1;
                    j += 1;
                }
            }
        }

        i == word.len() && j == abbr.len()
    }
}
